server/chat.ts

```diff
@@ -1141,6 +1116,53 @@ export class CommandContext extends MessageContext {
 
 		return message;
 	}
+	canPM(targetUser: User) {
+		const user = this.user;
+
+		// for hotpatching, remove later.
+		const curSettings = Chat.updateSettings(user); // will just be user.settings
+		const tarSettings = Chat.updateSettings(targetUser); // will just be targetUser.settings
+
+		if (targetUser.locked && !user.can('lock')) {
+			this.errorReply(`The user "${targetUser.name}" is locked and cannot be PMed.`);
+			return null;
+		}
+		if (
+			Config.pmmodchat && !Users.globalAuth.atLeast(user, Config.pmmodchat) &&
+			!Users.Auth.hasPermission(targetUser, 'promote', Config.pmmodchat as GroupSymbol)
+		) {
+			const groupName = Config.groups[Config.pmmodchat] && Config.groups[Config.pmmodchat].name || Config.pmmodchat;
+			this.errorReply(`On this server, you must be of rank ${groupName} or higher to PM users.`);
+			return null;
+		}
+		if (
+			(tarSettings.all === true ||
+			!Users.globalAuth.atLeast(user, tarSettings.all as GroupSymbol) && typeof curSettings.all === 'string' ||
+			tarSettings.specific?.includes(user.id)) &&
+			!user.can('lock')
+		) {
+			Chat.maybeNotifyBlocked('pm', targetUser, user);
+			if (!targetUser.can('lock')) {
+				this.errorReply(`This user is blocking private messages right now.`);
+				return null;
+			} else {
+				this.errorReply(`This ${Config.groups[targetUser.group].name} is too busy to answer private messages right now. Please contact a different staff member.`);
+				this.sendReply(`|html|If you need help, try opening a <a href="view-help-request" class="button">help ticket</a>`);
+				return null;
+			}
+		}
+		if (
+			(curSettings.all === true ||
+			!Users.globalAuth.atLeast(user, curSettings.all as GroupSymbol) && typeof tarSettings.all === 'string' ||
+			curSettings.specific?.includes(user.id)) &&
+			!targetUser.can('lock')
+		) {
+			const isBlocked = user.settings.blockPMs.specific?.includes(targetUser.id);
```

Comment:
Oh, this might be completely out of scope for this change (though I remembered while reading this), but one feature request I would super super super love is that if I as a global administrator/sysop who bypassesall/skips can etc PMs someone but my PMs are blocked so they can't respond, could you please at least give me a friendly warning that the user wont be able to respond? This would be huge for me

Reply:
I vaguely remember recommending making the check a method so it can be symmetric. I would much rather `bypassall` just not bypass PM permission checks asymmetrically.

Reply:
So should i not let admins bypass the "you are blocking pms and so can't pm x user check" then? 
(If i'm understanding correctly)

Reply:
Yes, you should not let admins bypass it at all.

Reply:
Okay.